doc/man/lipsum.1
.TH "LIPSUM" "1" "September 2014" "lipsum 0.1.15" "User Commands"
.SH "NAME"
lipsum -- Canonical example for the command module.
.SH "USAGE"

.SP
lipsum <command> [\-ljcveh] [\-\-color|\-\-no\-color] [\-\-debug]
.br
       [\-l|\-\-latin] [\-j|\-\-json] [\-c|\-\-collapse] [\-v|\-\-vanilla]
.br
       [\-e|\-\-exit] [\-h|\-\-help] [\-\-version] [\-\-log\-level=<level>]
.br
       [\-\-log\-file=<file>] [\-s|\-\-sort=(null|true|false|1|\-1)]
.br
       [\-f|\-\-format=(text|json|markdown|man)]
.br
       [\-a|\-\-align=(column|line|flex|wrap)] [\-m|\-\-maximum=<60\-240>]
.br
       <args>
.SH "DESCRIPTION"
.PP
Generates lorem ipsum text and is used to test help output for the command module.
.PP
If the \fB\-l | \-\-latin\fR option is specified and it looks like latin then it does nothing, other more meaningful options that have an affect are interspersed, try \fB\-\-align\fR and \fB\-\-format\fR in particular. The examples are valid, they illustrate some of the functionality of the command module.
.PP
When invoked without arguments the program will print some lorem ipsum paragraphs.
.SH "COMMANDS"
.BL
.IP "\[ci]" 4
\fBhelp\fR: Show help for commands.
.IP "\[ci]" 4
\fBprint, p\fR: Print some messages using the log middleware.
.IP "\[ci]" 4
\fBexception, ex, e\fR: Throw an exception.
.EL
.SH "OPTIONS"
.BL
.IP "\[ci]" 4
\fB\-v, \-\-vanilla\fR: Disable parameter replacement.
.IP "\[ci]" 4
\fB\-\-[no]\-color\fR: Enable or disable terminal colors.
.IP "\[ci]" 4
\fB\-\-log\-file=[file]\fR: Redirect to log file.
.IP "\[ci]" 4
\fB\-\-debug\fR: Enable stack traces.
.IP "\[ci]" 4
\fB\-l, \-\-latin\fR: Include mock lipsum options and commands.
.IP "\[ci]" 4
\fB\-j, \-\-json\fR: Print help as json.
.IP "\[ci]" 4
\fB\-c, \-\-collapse\fR: Collapse whitespace between sections.
.IP "\[ci]" 4
\fB\-\-log\-level=[level]\fR: Set the log level.
.IP "\[ci]" 4
\fB\-e, \-\-exit\fR: Include exit section from error definitions.
.IP "\[ci]" 4
\fB\-s, \-\-sort=[value]\fR: Alters the help option sort order. Set to null to use natural order which is likely the order that options were declared in however this is not guaranteed. Use false for the default sorting logic which favours options with more names, use true to sort lexicographically using \fIArray.prototype.sort\fR. Use 1 to sort by option string length (determined by the length of the help option string), reverse the order with \-1.
.IP "\[ci]" 4
\fB\-f, \-\-format=[value]\fR: Set the help output format.
.IP "\[ci]" 4
\fB\-a, \-\-align=[value]\fR: Alignment style.
.IP "\[ci]" 4
\fB\-m, \-\-maximum=[value]\fR: Maximum column width.
.IP "\[ci]" 4
\fB\-h, \-\-help\fR: Display this help and exit.
.IP "\[ci]" 4
\fB\-\-version\fR: Output version information and exit.
.EL
.SH "BUGS"
.PP
Report bugs to muji noop@xpm.io.


lib/lipsum.js
var util = require('util')
  , fs = require('fs')
  , path = require('path')
  , CommandInterface = require('cli-interface').CommandInterface
  , cli = require('cli-command')
  , help = require('cli-help')
  , midcolor = require('cli-mid-color')
  , midlogger = require('cli-mid-logger')
  , middebug = require('cli-mid-debug')
  , midmanual = require('cli-mid-manual')
  , convert = require('cli-native')
  , formats = help.fmt.formats
  , lipsum = '' + fs.readFileSync(path.join(__dirname, 'lipsum.txt'))
  , paragraphs = lipsum.split('\n\n')
  , list = [null, true, false, 1, -1]
  , alignments = ['column', 'line', 'flex', 'wrap']
  , stash = {};

// command handlers
function print(info, req, next) {
  var keys = require('cli-logger').keys;
  var log = this.log;
  keys.forEach(function(key) {
    log[key]('mock %s message', key);
  })
  next();
}

function exception(info, req, next) {
  next('%s command invoked', [info.cmd.getOptionString(' | ')]);
}

// add mock options and commands (--latin)
function mock() {
  this.option('-L, --lorem-ipsum-dolor[=VALUE]', 'Lorem ipsum dolor')
    .option('-i, --ipsum [VALUE]', 'Ipsum dolor sit amet')
    .option('--mauris-pulvinar, --ut-bibendum=[VALUE]', 'Mauris pulvinar')
    .option('-a, --aliquam [VALUE]', paragraphs[2])

  var program = this;
  paragraphs.forEach(function(para) {
    var cmd = para.replace(/\s+.*/, '').trim().toLowerCase();
    var desc = para.substr(0, para.indexOf('.') + 1);
    program.command(cmd, desc);
  })
}

// properties to associate with the commands
// and options loaded from lipsum.md
var options = {
  commands: {
    print: print,
    exception: exception
  },
  options: {
    sort: cli.types.enum(list, true),
    format: cli.types.enum(formats),
    align: cli.types.enum(alignments),
    maximum: cli.types.integer
  }
};


var LoremIpsum = function() {
  CommandInterface.apply(this, arguments);
}

util.inherits(LoremIpsum, CommandInterface);

LoremIpsum.prototype.configure = function() {
  var file = path.join(__dirname, 'lipsum.md');
  var conf = {
    trace: process.env.NODE_ENV === 'devel',
    compiler: {
      input: [file],
      definition: options
    },
    manual: {
      dir: path.normalize(path.join(__dirname, '..', 'doc', 'man')),
      dynamic: true
    }
  };
  var hints = {
    values: {
      format: formats,
      align: alignments,
      sort: list,
      maximum: [60, 240]
    }
  }
  this
    .configure(conf)
    //.usage(usage)
    //.usage(false)
    //.usage(null)
    .usage(hints);
}

LoremIpsum.prototype.use = function() {
  this
    .use(midcolor)
    .use(midlogger, null, {level: {}, file: {}})
    .use(middebug)
    .use(midmanual);
}

LoremIpsum.prototype.on = function() {
  this
    .once('load', function(req) {
      //console.log(this.name());
      this.help('-h, --help')
        .version();
    })
    .once('latin', function(req, arg, value) {
      mock.call(this);
    })
    .once('json', function(req, arg, value) {
      if(value) {
        this.configure().help.style = 'json';
        this.configure().help.indent = 2;
      }
    })
    .once('format', function(req, arg, value) {
      this.configure().help.style = value;
    })
    .once('maximum', function(req, arg, value) {
      this.configure().help.maximum = parseInt(value);
    })
    .once('exit', function(req, arg, value) {
      this.configure().help.exit = value;
    })
    .once('collapse', function(req, arg, value) {
      this.configure().help.collapse = value;
    })
    .once('vanilla', function(req, arg, value) {
      this.configure().help.vanilla = value;
    })
    .once('sort', function(req, arg, value) {
      this.configure().help.sort = convert.to(value);
    })
    .once('align', function(req, arg, value) {
      this.configure().help.align = value;
    })
    .on('empty', function(req) {
      process.stdout.write(lipsum);
    });
}

module.exports = function(pkg) {
  return new LoremIpsum(pkg);
}
